pkg/cmd/server/kubernetes/node_auth.go

```diff
@@ -0,0 +1,119 @@
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+
+	oauthenticator "github.com/openshift/origin/pkg/auth/authenticator"
+	"github.com/openshift/origin/pkg/auth/authenticator/request/bearertoken"
+	"github.com/openshift/origin/pkg/auth/authenticator/request/unionrequest"
+	"github.com/openshift/origin/pkg/auth/authenticator/request/x509request"
+	"github.com/openshift/origin/pkg/auth/authenticator/token/apitoken"
+	authncache "github.com/openshift/origin/pkg/auth/authenticator/token/cache"
+	"github.com/openshift/origin/pkg/auth/group"
+	authorizationapi "github.com/openshift/origin/pkg/authorization/api"
+	"github.com/openshift/origin/pkg/authorization/authorizer/adapter"
+	authzcache "github.com/openshift/origin/pkg/authorization/authorizer/cache"
+	"github.com/openshift/origin/pkg/authorization/authorizer/subjectaccessreview"
+	oclient "github.com/openshift/origin/pkg/client"
+	"github.com/openshift/origin/pkg/cmd/server/bootstrappolicy"
+)
+
+func newAuthenticator(clientCAs *x509.CertPool, anonymousConfig client.Config) (authenticator.Request, error) {
+	authenticators := []oauthenticator.Request{}
+
+	// API token auth
+	tokenAuthenticator, err := apitoken.NewTokenAuthenticator(anonymousConfig)
+	if err != nil {
+		return nil, err
+	}
+	// Cache results
+	cachedTokenAuthenticator, err := authncache.NewAuthenticator(tokenAuthenticator, 5*time.Minute, 1000)
+	if err != nil {
+		return nil, err
+	}
+	authenticators = append(authenticators, bearertoken.New(cachedTokenAuthenticator, true))
+
+	// Client-cert auth
+	if clientCAs != nil {
+		opts := x509request.DefaultVerifyOptions()
+		opts.Roots = clientCAs
+		certauth := x509request.New(opts, x509request.SubjectToUserConversion)
+		authenticators = append(authenticators, certauth)
+	}
+
+	ret := &unionrequest.Authenticator{
+		FailOnError: true,
+		Handlers: []oauthenticator.Request{
+			group.NewGroupAdder(unionrequest.NewUnionAuthentication(authenticators...), []string{bootstrappolicy.AuthenticatedGroup}),
+			// TODO: allow system:anonymous?
+			// authenticator.RequestFunc(func(req *http.Request) (user.Info, bool, error) {
+			// 	return &user.DefaultInfo{Name: bootstrappolicy.UnauthenticatedUsername, Groups: []string{bootstrappolicy.UnauthenticatedGroup}}, true, nil
+			// }),
+		},
+	}
+
+	return ret, nil
+}
+
+func newAuthorizerAttributesGetter() (authorizer.RequestAttributesGetter, error) {
+	return NodeAuthorizerAttributesGetter{}, nil
+}
+
+type NodeAuthorizerAttributesGetter struct{}
+
+func isSubpath(r *http.Request, path string) bool {
+	path = strings.TrimSuffix(path, "/")
+	return r.URL.Path == path || strings.HasPrefix(r.URL.Path, path+"/")
+}
+func (NodeAuthorizerAttributesGetter) GetRequestAttributes(u user.Info, r *http.Request) authorizer.Attributes {
+	attrs := authorizer.AttributesRecord{
+		User: u,
+		Verb: strings.ToLower(r.Method),
+	}
+
+	switch {
+	case isSubpath(r, "/stats"):
+		attrs.Resource = authorizationapi.NodeStatsResource
```

Comment:
If you're pretending this is a resource, use one of our resource verbs, not just the http method.